### PR TITLE
Subscriber to send all Agent events to Island

### LINKED ISSUE #2177: Push Events from Agent to Island
## Description

Write a subscriber in the agent that subscribes to all events and pushes them to the `/api/events` endpoint on the Island. The subscriber should receive events and push them to the Island asynchronously. To achieve this, the subscriber can push all events onto a queue and have a separate thread that waits on the queue and sends a sequence of events to the endpoint.

## Tasks
- [x] Create a subscriber that adds events to an internal queue (0d) - @shreyamalviya 
    - [x] Subscribe for all events
    - [x] Consider whether [this](https://github.com/guardicore/monkey/pull/2182#discussion_r941134053) is needed
- [ ] Add a thread to the subscriber that reads from the queue and sends a sequence of events to the endpoint. This should try to batch events using some strategy. (0.25d) - @shreyamalviya

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do?

Fixes #2177 

What's left:
- ~Serializing the event before sending it (should mostly be addressed in #2275, but may need tweaking since we need the event topic to be sent as well; see [this](https://github.com/guardicore/monkey/pull/2182#discussion_r941134053))~
- ~UTs~
- ~Docstrings~

## PR Checklist
* [x] Have you added an explanation of what your changes do and why you'd like to include them?
* [x] Is the TravisCI build passing?
* [ ] ~Was the CHANGELOG.md updated to reflect the changes?~
* [ ] ~Was the documentation framework updated to reflect the changes?~
* [x] Have you checked that you haven't introduced any duplicate code?

## Testing Checklist

* [x] Added relevant unit tests?
* [x] Have you successfully tested your changes locally? Elaborate:
    > Tested by running manually and running UTs
* [x] If applicable, add screenshots or log transcripts of the feature working
    > See comment for logs

### CODE REVIEW

File: monkey/infection_monkey/send_all_events_to_island.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
+import logging
+import queue
+import threading
+from typing import Set
+
+import requests
+
+# TODO: shouldn't leak implementation information; can we do this some other way?
+from pubsub import pub
```
Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:9]: You just don't need it. Firstly, `topic` in `__call__()` will always be "all_events_topic", so the log message isn't useful. Secondly, you can get the event type from the event itself with `type(event).__name__`.

File: monkey/infection_monkey/send_all_events_to_island.py [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+
+
+class send_all_events_to_island:
+    """
+    Sends information about the events carried out by the Agent to the Island in batches
+    """
+
+    def __init__(self, server_address: str):
+        self._server_address = server_address
+        self._queue: queue.Queue[AbstractAgentEvent] = queue.Queue()
```
Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:29]: Can Queue be typehinted with generic in python3.7? I feel like I've tried this before and it doesn't work.

File: monkey/infection_monkey/send_all_events_to_island.py [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+from pubsub import pub
+
+from common.common_consts.timeouts import MEDIUM_REQUEST_TIMEOUT
+from common.events import AbstractAgentEvent
+from common.utils import Timer
+
+logger = logging.getLogger(__name__)
+
+
+DEFAULT_TIME_PERIOD = 5
```
Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:18]: Add units to the variable name

File: monkey/infection_monkey/send_all_events_to_island.py [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+        """
+        Handles the batching and sending of the Agent's events to the Island
+        """
+
+        def __init__(self, queue_of_events: queue.Queue, server_address: str, time_period: int):
+            self._queue = queue_of_events
+            self._server_address = server_address
+            self._time_period = time_period
+
+            self._event_batch: Set = set()
```
Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:55]: What does the `Set` contain?

File: monkey/infection_monkey/send_all_events_to_island.py [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+            self._should_run_batch_and_send_thread = True
```
Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:56]: Use `threading.Event`.

File: monkey/infection_monkey/send_all_events_to_island.py [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+                        data=serialized_event,
+                        headers={"content-type": "application/json"},
+                        verify=False,
+                        timeout=MEDIUM_REQUEST_TIMEOUT,
+                    )
+                except Exception as exc:
+                    logger.warning(
+                        f"Exception caught when connecting to the Island at {self.server_address}"
+                        f": {exc}"
+                    )
```
Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:102]: This seems like it has extra steps in it.

```
while not thread_is_stopped:
    time.sleep(self._time_period)
    wake up
    read queue until it's empty
    send all events
````

Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:102]: This loop is still sending individual events. The point of batching them is to minimize traffic. This should be sending a sequence of serialized events.

File: monkey/infection_monkey/send_all_events_to_island.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                self._event_batch.add(event)
+            except queue.Empty:
+                pass
+
+        def _send_events_to_island(self):
+            for serialized_event in self._event_batch:
+                try:
+                    requests.post(  # noqa: DUO123
+                        "https://%s/api/events" % (self._server_address,),
+                        data=serialized_event,
```
Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:93]: Use `json=`

File: monkey/infection_monkey/send_all_events_to_island.py [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+        def _send_remaining_events(self):
+            while not self._queue.empty():
+                self._add_next_event_to_batch()
+
+            self._send_events_to_island()
+
+        def stop(self):
+            self._should_run_batch_and_send_thread = False
+            self._batch_and_send_thread.join()
```
Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:112]: We could make this class a subclass of `Thread` and use `InterruptableThreadMixin` after `2216-tcp-relay` is merged.

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/infection_monkey/send_all_events_to_island.py (hunk 2 above)
Comment (shreyamalviya) [reply monkey/infection_monkey/send_all_events_to_island.py:25]: What's wrong with doing it this way? It's how we do it in other places too.

**Review by ilija-lazoroski — APPROVED** [review]

File: monkey/infection_monkey/send_all_events_to_island.py (hunk 2 above)
Comment (mssalvatore) [reply monkey/infection_monkey/send_all_events_to_island.py]: Nothing if it works. I'm just saying I've tried it before and had issues. Maybe `typing.Queue` isn't generic and that's what I tried to use.

File: monkey/infection_monkey/send_all_events_to_island.py [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+        self._server_address = server_address
+        self._time_period = time_period
+
+        self._queue: queue.Queue[AbstractAgentEvent] = queue.Queue()
+        self._stop_batch_and_send_thread = threading.Event()
+
+    def start(self):
+        self._batch_and_send_thread = threading.Thread(
+            name="SendEventsToIslandInBatchesThread", target=self._manage_event_batches
+        )
```
Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:64]: use `create_daemon_thread()`

File: monkey/infection_monkey/send_all_events_to_island.py [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                EVENTS_API_URL % (self._server_address,),
+                json=events,
+                verify=False,
+                timeout=MEDIUM_REQUEST_TIMEOUT,
+            )
+        except Exception as exc:
+            logger.warning(
+                f"Exception caught when connecting to the Island at {self._server_address}"
+                f": {exc}"
+            )
```
Comment (mssalvatore) [inline monkey/infection_monkey/send_all_events_to_island.py:97]: Let's log the sequence of events we're sending. I think something like `logger.debug(f"sending: {events})` should work.

File: monkey/tests/unit_tests/infection_monkey/test_send_all_events_to_island.py [hunk 11]
```diff
@@ -0,0 +1,47 @@
...
+
+def test_send_events(event_sender):
+    with requests_mock.Mocker() as mock:
+        mock.post(EVENTS_API_URL % SERVER)
+
+        event_sender.start()
+
+        for _ in range(5):
+            event_sender.add_event_to_queue({})
+        time.sleep(0.1)
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/infection_monkey/test_send_all_events_to_island.py:24]:
```suggestion
        time.sleep(0.005)
```

File: monkey/tests/unit_tests/infection_monkey/test_send_all_events_to_island.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
...
+        assert mock.call_count == 1
+
+        event_sender.add_event_to_queue({})
+        time.sleep(0.1)
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/infection_monkey/test_send_all_events_to_island.py:28]:
```suggestion
        time.sleep(0.005)
```

File: monkey/tests/unit_tests/infection_monkey/test_send_all_events_to_island.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+def test_send_remaining_events(event_sender):
+    with requests_mock.Mocker() as mock:
+        mock.post(EVENTS_API_URL % SERVER)
+
+        event_sender.start()
+
+        for _ in range(5):
+            event_sender.add_event_to_queue({})
+        time.sleep(0.1)
```
Comment (mssalvatore) [inline monkey/tests/unit_tests/infection_monkey/test_send_all_events_to_island.py:42]:
```suggestion
        time.sleep(0.005)
```

**Review by mssalvatore — CHANGES_REQUESTED** [review]

File: monkey/tests/unit_tests/infection_monkey/test_send_all_events_to_island.py (hunk 13 above)
Comment (shreyamalviya) [reply monkey/tests/unit_tests/infection_monkey/test_send_all_events_to_island.py:42]: 0.005 is too less. The test fails. Reduced it to 0.5.

Comment (mssalvatore) [reply monkey/tests/unit_tests/infection_monkey/test_send_all_events_to_island.py]: What about 0.01? That's 10x the sleep period, which should be enough.

File: monkey/infection_monkey/agent_event_forwarder.py [hunk 14]
```diff
@@ -0,0 +1,106 @@
...
+    Sends information about the events carried out by the Agent to the Island in batches
+    """
+
+    def __init__(
+        self, server_address: str, agent_event_serializer_registry: EventSerializerRegistry
+    ):
+        self._server_address = server_address
+        self._agent_event_serializer_registry = agent_event_serializer_registry
+
+        self._agent_events_to_island_sender = BatchingAgentEventForwarder(self._server_address)
```
Comment (mssalvatore) [inline monkey/infection_monkey/agent_event_forwarder.py:32]:
```suggestion
        self._batching_agent_event_forwarder = BatchingAgentEventForwarder(self._server_address)
```

**Pull request merged**